core/src/test/java/hudson/scheduler/CronTabTest.java

```diff
@@ -314,4 +314,46 @@ public int next(int n) {
         assertEquals("[35, 56]", times.toString());
     }
 
+    @Issue("JENKINS-18313")
+    @Test public void passingMidnight() throws Exception {
+        // Every Hour
+        CronTabList tabs = CronTabList.create("0 20-5 * * *", Hash.from("seed"));
+        List<Integer> times = new ArrayList<Integer>();
+        for (int i = 0; i < 24; i++) {
+            if (tabs.check(new GregorianCalendar(2013, 3, 3, i, 0, 0))) {
+                times.add(i);
+            }
+        }
+        assertEquals("[0, 1, 2, 3, 4, 5, 20, 21, 22, 23]", times.toString());
```

Comment:
None of the tests check which actual _date_ the hours are on.

My expectation is that e.g. `H H(21-6) * * 0` is expected to be defined as a random time from a continuous 9 hour period starting Sunday 9 PM and ending Monday 6 AM. I wouldn't be surprised if this implementation would be "either Sunday morning or Sunday evening".

Of course, the complete lack of documentation doesn't help.

Reply:
There is no check for the date, since the input assumes daily.

```H H(21-6) * * 0``` will be a random point either Sunday morning between 0 and 6 or in the evening between 21 and 24 (23:59 as per original code). The question is if it might make sense to define it like this, I only checked the hours without dates, because my use-case is a daily run. But I can add some test and documentation if it helps get this merged.

Reply:
> The question is if it might make sense to define it like this, I only checked the hours without dates, because my use-case is a daily run.

Thanks for the clarification! I asked for feedback in an issue comment to see what others think about this.